executor/pipeline/modifiers.rs

```diff
@@ -0,0 +1,373 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+use std::{cmp::Ordering, collections::HashMap, marker::PhantomData, sync::Arc};
+
+use compiler::{
+    modifiers::{FilterProgram, LimitProgram, OffsetProgram, SortProgram},
+    VariablePosition,
+};
+use ir::program::modifier::SortVariable;
+use lending_iterator::LendingIterator;
+use storage::snapshot::ReadableSnapshot;
+use crate::{
+    batch::Batch,
+    ExecutionInterrupt,
+    pipeline::{PipelineExecutionError, StageAPI, StageIterator},
+    row::MaybeOwnedRow,
+};
+
+// Sort
+pub struct SortStageExecutor<Snapshot, PreviousStage> {
+    program: SortProgram,
+    previous: PreviousStage,
+    phantom: PhantomData<Snapshot>,
+}
+
+impl<Snapshot, PreviousStage> SortStageExecutor<Snapshot, PreviousStage>
+where
+    Snapshot: ReadableSnapshot + 'static,
+    PreviousStage: StageAPI<Snapshot>,
+{
+    pub fn new(program: SortProgram, previous: PreviousStage) -> Self {
+        Self { program, previous, phantom: PhantomData::default() }
+    }
+}
+
+impl<Snapshot, PreviousStage> StageAPI<Snapshot> for SortStageExecutor<Snapshot, PreviousStage>
+where
+    Snapshot: ReadableSnapshot + 'static,
+    PreviousStage: StageAPI<Snapshot>,
+{
+    type OutputIterator = SortStageIterator<Snapshot>;
+
+    fn named_selected_outputs(&self) -> HashMap<VariablePosition, String> {
```

Comment:
We can eliminate this by just exposing `selected_output_mapping`, which maps `Map<VariablePosition, Variable>`. (probably inverted actually, see next)

Then in the `compile` of the pipeline, we can retain the Variable registry, which goes from `String` variable name to `Variable`. Using these two maps we should be able to reconstruct `Map<VariablePosition, String>` only in the top compilation and return that alongside the pipeline... or create a Pipeline construct in `//query` which holds exactly the last stage and this map!